src/Microsoft.AspNetCore.SignalR/HubEndPoint.cs

```diff
@@ -295,6 +307,68 @@ public class HubEndPoint<THub, TClient> : EndPoint, IInvocationBinder where THub
             }
         }
 
+        private bool IsChannel(Type type, out Type payloadType)
+        {
+            var channelType = type.AllBaseTypes().FirstOrDefault(t => t.IsGenericType && t.GetGenericTypeDefinition() == typeof(ReadableChannel<>));
+            if (channelType == null)
+            {
+                payloadType = null;
+                return false;
+            }
+            else
+            {
+                payloadType = channelType.GetGenericArguments()[0];
+                return true;
+            }
+        }
+
+        private async Task StreamResultsAsync(string invocationId, Connection connection, IHubProtocol protocol, IAsyncEnumerator<object> enumerator)
+        {
+            // TODO: Cancellation?
+            try
+            {
+                while (await enumerator.MoveNextAsync())
+                {
+                    // Send the stream item
+                    await SendMessageAsync(connection, protocol, new StreamItemMessage(invocationId, enumerator.Current));
+                }
+
+                await SendMessageAsync(connection, protocol, CompletionMessage.Empty(invocationId));
+            }
+            catch (Exception ex)
+            {
+                await SendMessageAsync(connection, protocol, CompletionMessage.WithError(invocationId, ex.Message));
+            }
+        }
+
+        private bool IsStreamed(ObjectMethodExecutor methodExecutor, object result, out IAsyncEnumerator<object> enumerator)
+        {
+            // TODO: Cache attributes the OME?
+            var streamingAttribute = methodExecutor.MethodInfo.ReturnParameter.GetCustomAttribute<StreamingAttribute>();
+            if (streamingAttribute == null)
+            {
+                enumerator = null;
+                return false;
+            }
+
+            var observableInterface = result.GetType().GetInterfaces().FirstOrDefault(i => i.IsGenericType && i.GetGenericTypeDefinition() == typeof(IObservable<>));
+            if (observableInterface != null)
+            {
+                enumerator = AsyncEnumeratorAdapters.FromObservable(result, observableInterface);
+                return true;
+            }
+            else if (IsChannel(result.GetType(), out var payloadType))
+            {
+                enumerator = AsyncEnumeratorAdapters.FromChannel(result, payloadType);
+                return true;
+            }
+            else
+            {
+                // This type cannot be streamed
+                throw new NotSupportedException($"Cannot stream results of type: {result.GetType().FullName}");
```

Comment:
If we keep the attribute, sure. But if we remove the attribute then single results just won't be streamed.